bark_ml/environments/single_agent_runtime.py

```diff
@@ -61,6 +61,8 @@ def step(self, action):
 
     # step and observe
     self._world.Step(self._step_time)
+    if eval_id not in self._world.agents:
+      return [0] * self._observer.observation_space.shape[0], -1, True, None
     observed_world = self._world.Observe([eval_id])[0]
```

Comment:
Adding this change was necessary to train the SAC, PPO and GAIL agents on the highway and intersection blueprint. After the `Step()` function sometimes the eval agent is not valid anymore and thus the `Observe()` method raises an IndexError. 
We decided to fix it by returning a mock observation, negative reward, done flag set and no further environment info.

Reply:
If the agent is not valid anymore, why is the episode not terminal?
Additionally, there is a flag: `params["World"]["remove_agents_out_of_map"] = True`

Reply:
![image](https://user-images.githubusercontent.com/17546974/90146960-f63e0200-dd81-11ea-8538-655b5c94d23f.png)

This is what happens in the `examples/tfa.py` script (No changes from our side) when using the `ContinuousHighwayBlueprint`. Setting `params["World"]["remove_agents_out_of_map"] = False` prevents the error.

Is this intended behaviour and shall we remove the check if the agent is valid?

Reply:
The fix as it is currently implemented is not a viable option as this would only work for the discrete behavior model.

I would suggest to add a log message:
```python
if eval_id not in world.agents:
  logger.info(f"Eval Agent-ID {eval_id} does not exist in the current world.")
```

And the flag should be additionally be set to true.

Reply:
I removed the lines completely, as when we use the `params["World"]["remove_agents_out_of_map"] = False` flag the agent is always valid and the logging is never triggered. If we use `params["World"]["remove_agents_out_of_map"] = True` the logging is triggered, but the next line of code triggeres the IndexError and stops the execution.